core/src/main/java/org/elasticsearch/action/get/MultiGetResponse.java

```diff
@@ -149,6 +186,56 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+        try {
+            String currentFieldName = null;
+            for (Token token = parser.nextToken(); token != Token.END_OBJECT; token = parser.nextToken()) {
+                switch (token) {
+                    case FIELD_NAME:
+                        currentFieldName = parser.currentName();
+                        break;
+                    case START_ARRAY:
+                        if (Fields.DOCS.equals(currentFieldName) == false) {
+                            throw new ElasticsearchParseException("Unexpected field [" + currentFieldName + "]");
+                        }
+                        for (Token innerToken = parser.nextToken(); innerToken != Token.END_ARRAY; innerToken = parser.nextToken()) {
+                            if (innerToken == Token.START_OBJECT) {
+
+                                // Parsing this in a streaming manner means that at least up to 4 fields need be
+                                // keep tracked of here and then GetResult.fromXContent(...) needs to be broken up.
+                                // This is makes the parsing code complex, whereas reading each json object into a map and
+                                // then creating a parser from that is a bit inefficient, but makes the parsing code less
+                                // complex. I think this is the right trade off. Unless we introduce a new xcontent format
+                                // in mget response for the hl rest client that is enabled via a parameter that renders
+                                // an unambiguous format.
+                                XContentType contentType = parser.contentType();
+                                Map<String, Object> map = parser.map();
+                                XContentBuilder builder = XContentBuilder.builder(contentType.xContent());
+                                builder.map(map);
+                                BytesReference bytes = builder.bytes();
+                                XContentParser parser2 = XContentHelper.createParser(parser.getXContentRegistry(), bytes, contentType);
+
+                                if (map.containsKey("error")) {
+                                    Failure failure = Failure.fromXContent(parser2);
+                                    items.add(new MultiGetItemResponse(null, failure));
+                                } else {
+                                    GetResponse getResponse = GetResponse.fromXContent(parser2);
+                                    items.add(new MultiGetItemResponse(getResponse, null));
+                                }
+                            }  else {
+                                throw new ElasticsearchParseException("Unexpected token [" + innerToken + "]");
+                            }
+                        }
+                        break;
+                }
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
```

Comment:
why can't fromXContent throw IOException? We do this everywhere I think.

Reply:
good point, let me change that